src/main/groovy/org/boozallen/plugins/jte/console/TemplateLogger.groovy

```diff
@@ -0,0 +1,82 @@
+package org.boozallen.plugins.jte.console
+
+import org.boozallen.plugins.jte.Utils
+import hudson.Extension
+import hudson.MarkupText
+import hudson.Util
+import hudson.console.ConsoleAnnotationDescriptor
+import hudson.console.ConsoleAnnotator
+import hudson.console.ConsoleNote
+import hudson.model.TaskListener
+import org.jenkinsci.plugins.workflow.job.WorkflowRun
+
+enum LogLevel{
+    INFO(tag: "info"),
+    WARN(tag: "warn"),
+    ERROR(tag: "error")
+    String tag
+}
+
+public class TemplateLogger extends ConsoleNote<WorkflowRun> {
+
+    private static final String CONSOLE_NOTE_PREFIX = "[JTE] "
+    String messageID 
+    LogLevel logType 
+    Boolean firstLine
+    Boolean multiLine
+    Boolean initiallyHidden
+
+    @Override
+    ConsoleAnnotator<?> annotate(WorkflowRun context, MarkupText text, int charPos) {
+        def tags = [
+            "class='jte-${logType.tag}'",
+            "jte-id='${messageID}'",
+        ]
+        if(multiLine){
+            tags.push("first-line='${firstLine}''")
+            tags.push("multi-line='${multiLine}'")
+        }
+        if(initiallyHidden){
+            tags.push("initially-hidden='${initiallyHidden}'")
+        }
+        text.wrapBy("<span ${tags.join(" ")}>", '</span>')
+        return null
+    }
+
+    static void print(String message, Boolean initiallyHidden = false, LogLevel logType = LogLevel.INFO, Boolean trimLines = true) {
```

Comment:
getting to the point where we should probably split these input arguments into required and optional? 

here's some neat code to validate the input args if we switch to a map: 
```
static void print(String message, Map config){
  argTypes = [ 
    "initiallyHidden": Boolean, 
    "logType": LogLevel, 
    "trimLines": Boolean
  ]

  Boolean throwError = false 
  String errorMsg = "Improper TemplateLogger Configuration: \n"
  
def improperConfigNames = config.keySet() - argTypes.keySet()
  if(improperConfigNames){
      throwError = true 
      errorMsg +=  "The following keys are not identified: ${improperConfigNames} \n"
      improperConfigNames.each{ config.remove(it) }
  }
  config.each{ key, value -> 
      Class expectedType = argTypes[key]
      if(!(value.getClass() == expectedType)){
          throwError = true 
          errorMsg += "key ${key} should be of type ${argTypes[key]} \n"
      }
  }
  
  if (throwError){
      throw new TemplateException(errorMsg)
  }
  
  // apply defaults if not set. 
  config = [ initiallyHidden: false, trimLines: true, logType: LogLevel.INFO ] + config 
  

   // do rest of the things..
   
}
```

then - the call to TemplateLogger can be easier to read: 

`TemplateLogger.print( "my message", [initiallyHidden: true, logType: LogLevel.WARN])`
`TemplateLogger.print("my message")` 
`TemplateLogger.print("my message", [trimLines: false]) `

Reply:
good idea, the required should probably be arguments, message and logType, the others may be best in the map at the end ... so probably: TemplateLogger.print(String s, LogLevel l = LogLevel.INFO, opts = [:]) ... then the code above or no default on LogLevel

Reply:
implemented